driver/csiplugin/nodeserver.go

```diff
@@ -234,6 +235,52 @@ func (ns *ScaleNodeServer) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfo
 func (ns *ScaleNodeServer) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolumeRequest) (*csi.NodeExpandVolumeResponse, error) {
 	return nil, status.Error(codes.Unimplemented, "")
 }
+
 func (ns *ScaleNodeServer) NodeGetVolumeStats(ctx context.Context, req *csi.NodeGetVolumeStatsRequest) (*csi.NodeGetVolumeStatsResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	glog.V(4).Infof("NodeGetVolumeStats called with req: %#v", req)
+
+	if len(req.VolumeId) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Volume ID must be provided")
+	}
+	if len(req.VolumePath) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats Target Path must be provided")
+	}
+
+	if _, err := os.Lstat(req.VolumePath); err != nil {
+		if os.IsNotExist(err) {
+			return nil, status.Errorf(codes.NotFound, "path %s does not exist", req.VolumePath)
+		}
+		return nil, status.Errorf(codes.Internal, "failed to stat path %s: %v", req.VolumePath, err)
+	}
+
+	volumeIDMembers, err := getVolIDMembers(req.GetVolumeId())
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, "NodeGetVolumeStats : VolumeID is not in proper format")
+	}
+
+	if !volumeIDMembers.IsFilesetBased {
+		return nil, status.Error(codes.InvalidArgument, "volume stat not supported for lightweight volumes")
+	}
+
+	available, capacity, used, inodes, inodesFree, inodesUsed, err := utils.FsStatInfo(req.GetVolumePath())
```

Comment:
Instead of FsStatInfo(), one can get similar info by calling GetFilesetQuotaDetails().
I know in this code file, calls for GPFS REST APIs are node made, but some way to use GPFS REST APIs cannot be used ?

Reply:
@madhuthorat This API is called by Kubernetes every 2 minutes and for each volume. The stat reports the same data as of gpfs api call so we can save all those api request by doing stat locally.